includes/payment-methods/class-wc-stripe-upe-payment-gateway.php

```diff
@@ -703,32 +703,31 @@ private function process_payment_with_deferred_intent( int $order_id ) {
 
 				// Create a payment intent, or update an existing one associated with the order.
 				$payment_intent = $this->process_payment_intent_for_order( $order, $payment_information );
-
-				// Handle saving the payment method in the store.
-				// It's already attached to the Stripe customer at this point.
-				if ( $payment_information['save_payment_method_to_store'] ) {
-					$this->handle_saving_payment_method(
-						$order,
-						$payment_information['payment_method'],
-						$selected_payment_type
-					);
-				}
-
-				// Use the last charge within the intent to proceed.
-				$charge = end( $payment_intent->charges->data );
-
-				// Only process the response if it contains a charge object. Intents with no charge require further action like 3DS and will be processed later.
-				if ( $charge ) {
-					$this->process_response( $charge, $order );
-				}
-
-				// Set the selected UPE payment method type title in the WC order.
-				$this->set_payment_method_title_for_order( $order, $selected_payment_type );
 			} else {
-				// It's a setup intent. To be handled.
-				return [ 'result' => 'failure' ];
+				$payment_intent = $this->process_setup_intent_for_order( $order, $payment_information );
```

Comment:
In the existing `process_payment()` method, we have the following code which handles the 'change_payment_method' flow separately:

```
if ( $this->maybe_change_subscription_payment_method( $order_id ) ) {
	return $this->process_change_subscription_payment_method( $order_id );
}
```

While working on this PR, I couldn't find a need to handle these requests separately. Am I missing something here?

Reply:
Hmm actually, looking at the `process_change_subscription_payment_method()` there are some differences:

- It throws a `do_action( 'wc_stripe_change_subs_payment_method_success' );` before returning
  - I can't see this hook being used by anything, but it's technically a breaking change so we can add it back.
- It doesn't call `$order->payment_complete()`
  - Calling `WC_Subscription::payment_complete()` exits on the first line if the request is to change a payment method so this is already being handled by Woo Subscriptions. I think it's better to still not even call payment complete while we're handling a change payment method request, so I'll update this PR.

I can't see the hook being used anyway

Reply:
I've just pushed up 10591299491085df343f37f2211d00c24d7c1490 which fixes the two issues above by:

- making sure `do_action( 'wc_stripe_change_subs_payment_method_success' )` is triggered inside process_payment_with_deferred_intent()
- don't call `$order->payment_complete()` when changing a subscriptions payment method.